### PR TITLE
LIBBEAT: Enhancement Convert dissected values from String to other basic data types and IP

### PR DESCRIPTION
**Pull request by premendrasingh**
## What does this PR do?
This PR enhances dissect processor to convert string values to integer, long, float, double, boolean or IP. Key and convert data type are separated by `|` separator.

## Why is it important?
It will save CPU cycles converting data from string to other data type using Convert processor

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally
Added unit and benchmark test cases.

## Related issues

Closes https://github.com/elastic/dissect-specification/issues/10

### CODE REVIEW

**Comment by jsoriano** [comment]
Nice enhancement!
I have a question about the way the type is specified. Is there any special reason to use the `|` separator?
[In Grok](https://www.elastic.co/guide/en/elasticsearch/reference/7.7/grok-processor.html#grok-basics) the `:` separator is used for the same purpose. And in the logstash implementation of dissect, the types are declared with an additional [`convert_datatype` option](https://www.elastic.co/guide/en/logstash/7.7/plugins-filters-dissect.html#plugins-filters-dissect-convert_datatype).
If there is no special reason to use the `|` separator, I think we should use one of the already existing methods to declare types.

**Comment by premendrasingh** [comment]
> Nice enhancement!
> I have a question about the way the type is specified. Is there any special reason to use the `|` separator?
> [In Grok](https://www.elastic.co/guide/en/elasticsearch/reference/7.7/grok-processor.html#grok-basics) the `:` separator is used for the same purpose. And in the logstash implementation of dissect, the types are declared with an additional [`convert_datatype` option](https://www.elastic.co/guide/en/logstash/7.7/plugins-filters-dissect.html#plugins-filters-dissect-convert_datatype).
> If there is no special reason to use the `|` separator, I think we should use one of the already existing methods to declare types.

`|` was chosen based on discussion in Issue #10 https://github.com/elastic/dissect-specification/issues/10
Please let me know whether I should change it to `:` (colon).

I think annotating key with data type right next to it, is more readable compared to [`convert_datatype`](https://www.elastic.co/guide/en/logstash/7.7/plugins-filters-dissect.html#plugins-filters-dissect-convert_datatype)  option.

Thanks for reviewing.

**Comment by sudeepknair** [comment]
Please share the timeline for review and merge for this change? We are looking forward to using this from the upstream code.

**Comment by sudeepknair** [comment]
any updates on this PR request?

File: CHANGELOG.next.asciidoc [hunk 1]
```diff
@@ -267,6 +267,8 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - When using the `decode_json_fields` processor, decoded fields are now deep-merged into existing event. {pull}17958[17958]
 - Add backoff configuration options for the Kafka output. {issue}16777[16777] {pull}17808[17808]
 - Add TLS support to Kerberos authentication in Elasticsearch. {pull}18607[18607]
+- Change ownership of files in docker images so they can be used in secured environments. {pull}12905[12905]
```
Comment (jsoriano) [inline CHANGELOG.next.asciidoc:270]: Not related to this change, please remove.

File: libbeat/processors/dissect/field.go [hunk 2]
```diff
@@ -161,7 +242,7 @@ type indirectField struct {
 func (f indirectField) Apply(b string, m Map) {
 	v, ok := m[f.Key()]
 	if ok {
-		m[v] = b
+		m[v.(string)] = b
```
Comment (jsoriano) [inline libbeat/processors/dissect/field.go:245]: For safety, to avoid panics, I think we should do some type check here. Maybe an option is to do nothing if the value is not a string.
```suggestion
		if v, ok := v.(string); ok {
			m[v] = b
		}
```

File: libbeat/processors/dissect/field.go [hunk 3]
```diff
@@ -186,7 +267,7 @@ type appendField struct {
 func (f appendField) Apply(b string, m Map) {
 	v, ok := m[f.Key()]
 	if ok {
-		m[f.Key()] = v + f.JoinString() + b
+		m[f.Key()] = v.(string) + f.JoinString() + b
```
Comment (jsoriano) [inline libbeat/processors/dissect/field.go:270]: Also here we should do something safer to avoid panics if `v` is not a string.

File: libbeat/processors/dissect/field.go [hunk 4]
```diff
@@ -90,8 +123,56 @@ type normalField struct {
...
-	m[f.Key()] = b
+	if len(f.dataType) == 0 {
+		m[f.Key()] = b
+	} else {
+		if dt, ok := dataTypeNames[f.dataType]; ok {
+			value, err := transformType(dt, b)
+			if err == nil {
+				m[f.Key()] = value
+			} else {
+				errors.Errorf("%s\n", err)
```
Comment (jsoriano) [inline libbeat/processors/dissect/field.go:172]: This line does nothing?

File: libbeat/processors/dissect/field.go [hunk 5]
```diff
@@ -90,8 +123,56 @@ type normalField struct {
...
+		return value, nil
+	}
+}
+
 func (f normalField) Apply(b string, m Map) {
-	m[f.Key()] = b
+	if len(f.dataType) == 0 {
+		m[f.Key()] = b
+	} else {
+		if dt, ok := dataTypeNames[f.dataType]; ok {
```
Comment (jsoriano) [inline libbeat/processors/dissect/field.go:167]: We should probably handle at some point the case of an unknown type.

File: libbeat/processors/dissect/field.go [hunk 6]
```diff
@@ -273,6 +354,21 @@ func newIndirectField(id int, key string, length int) indirectField {
 }
 
 func newNormalField(id int, key string, ordinal int, length int, greedy bool) normalField {
+	parts := strings.Split(key, "|")
```
Comment (jsoriano) [inline libbeat/processors/dissect/field.go:357]: Should the extraction of the type be responsibility of `extractKeyParts`?

And maybe `newField` should take care of checking that the type is a valid one.

File: libbeat/processors/dissect/config_test.go [hunk 7]
```diff
@@ -101,3 +101,51 @@ func TestConfig(t *testing.T) {
...
+		if !assert.NoError(t, err) {
+			return
+		}
+	})
+	t.Run("invalid data type", func(t *testing.T) {
+		c, err := common.NewConfigFrom(map[string]interface{}{
+			"tokenizer": "%{value1|int} %{value2|short} %{value3|char} %{value4|void} %{value5|unsigned}",
+			"field":     "message",
+		})
+		if !assert.NoError(t, err) {
```
Comment (jsoriano) [inline libbeat/processors/dissect/config_test.go:126]: I think we should expect an error on this case.

File: libbeat/processors/dissect/config_test.go [hunk 8]
```diff
@@ -101,3 +101,51 @@ func TestConfig(t *testing.T) {
...
+			return
+		}
+	})
+	t.Run("missing data type", func(t *testing.T) {
+		c, err := common.NewConfigFrom(map[string]interface{}{
+			"tokenizer": "%{value1|} %{value2|} %{value3|} %{value4|} %{value5|}",
+			"field":     "message",
+		})
+		if !assert.NoError(t, err) {
```
Comment (jsoriano) [inline libbeat/processors/dissect/config_test.go:141]: Also here, I think we should report an error.

File: libbeat/processors/dissect/dissect_test.go [hunk 9]
```diff
@@ -33,6 +33,83 @@ func TestNoToken(t *testing.T) {
...
+		{
+			Name: "Invalid data type, drop fields",
+			Tok:  "id=%{id|xyz} status=%{status|abc} msg=\"%{message}\"",
+			Msg:  "id=1945 status=500 msg=\"Internal server error\"}",
+			Expected: map[string]interface{}{
+				"message": "Internal server error",
+			},
+			Fail: false,
+		},
+	}
```
Comment (jsoriano) [inline libbeat/processors/dissect/dissect_test.go:88]: I think these cases should be added to `libbeat/processors/dissect/testdata/dissect_tests.json` instead.

Could you also please add some test case where some suffix like `->` is also used?

File: libbeat/processors/dissect/field.go [hunk 10]
```diff
@@ -90,8 +123,56 @@ type normalField struct {
...
+		// strconv.ParseInt will accept the '0x' or '0X` prefix only when base is 0.
+		base = 0
+	}
+	return strconv.ParseInt(s, base, bitSize)
+}
+
+func transformType(typ dataType, value string) (interface{}, error) {
+	switch typ {
+	case String:
+		return fmt.Sprintf("%v", value), nil
```
Comment (jsoriano) [inline libbeat/processors/dissect/field.go:139]: I think this conversion is not needed, the value is already a string.
```suggestion
		return value, nil
```

File: CHANGELOG.next.asciidoc (hunk 1 above)
Comment (premendrasingh) [reply CHANGELOG.next.asciidoc]: Removed this line in latest commit.

File: libbeat/processors/dissect/field.go (hunk 2 above)
Comment (premendrasingh) [reply libbeat/processors/dissect/field.go]: Updated to use code above.

File: libbeat/processors/dissect/field.go (hunk 3 above)
Comment (premendrasingh) [reply libbeat/processors/dissect/field.go]: Applied same logic as above.

File: libbeat/processors/dissect/field.go (hunk 4 above)
Comment (premendrasingh) [reply libbeat/processors/dissect/field.go]: Removed this line.

File: libbeat/processors/dissect/field.go (hunk 6 above)
Comment (premendrasingh) [reply libbeat/processors/dissect/field.go]: Refactored as suggested above.

Comment (premendrasingh) [reply libbeat/processors/dissect/field.go]: Extraction of type moved to extractKeyParts, new field checks for valid data type and returns error for invalid, missing data types.

File: libbeat/processors/dissect/config_test.go (hunk 7 above)
Comment (premendrasingh) [reply libbeat/processors/dissect/config_test.go:126]: Fixed code to generate error and expect error here.

File: libbeat/processors/dissect/config_test.go (hunk 8 above)
Comment (premendrasingh) [reply libbeat/processors/dissect/config_test.go:141]: Updated to expect error for missing data type.

File: libbeat/processors/dissect/dissect_test.go (hunk 9 above)
Comment (premendrasingh) [reply libbeat/processors/dissect/dissect_test.go:80]: Added test case for padding ->
Json handles only **number** data type. We need to test for integer, long, float data types also, which is not possible in Json. That's why I could not add test cases to dissect_tests.json and added them to dissect_tests.go instead.

File: libbeat/processors/dissect/field.go (hunk 10 above)
Comment (premendrasingh) [reply libbeat/processors/dissect/field.go]: Removed the redundant conversion.

**Comment by premendrasingh** [comment]
@jsoriano Thanks for the great feedback. I have committed the changes are suggested. Please review.

File: libbeat/processors/dissect/dissect_test.go [hunk 11]
```diff
@@ -66,22 +66,14 @@ func TestDissectConversion(t *testing.T) {
 			Fail: false,
 		},
 		{
-			Name: "Missing data type, drop fields",
-			Tok:  "id=%{id|} status=%{status|} msg=\"%{message}\"",
-			Msg:  "id=1857 status=404 msg=\"File not found\"}",
+			Name: "Greedy padding skip test ->",
+			Tok:  "id=%{id|integer->} padding_removed=%{padding_removed|boolean->} length=%{length|long->} msg=\"%{message}\"",
```
Comment (jsoriano) [inline libbeat/processors/dissect/dissect_test.go:70]: If I understand correctly the proposal in https://github.com/elastic/dissect-specification/issues/10#issuecomment-616464555, the type should be after the key and its suffixes, so this should be like this:
```suggestion
			Tok:  "id=%{id->|integer} padding_removed=%{padding_removed->|boolean} length=%{length->|long} msg=\"%{message}\"",
```

File: libbeat/processors/dissect/field.go [hunk 12]
```diff
@@ -353,36 +370,35 @@ func newIndirectField(id int, key string, length int) indirectField {
...
-func extractKeyParts(rawKey string) (key string, ordinal int, length int, greedy bool) {
+func extractKeyParts(rawKey string) (key string, dataType string, ordinal int, length int, greedy bool) {
 	m := suffixRE.FindAllStringSubmatch(rawKey, -1)
 
+	if m[0][1] != "" {
+		parts := strings.Split(m[0][1], dataTypeSeparator)
+		if len(parts) > 1 {
+			key = parts[0]
+			if parts[1] == "" {
+				dataType = "[unset]"
```
Comment (jsoriano) [inline libbeat/processors/dissect/field.go:394]: Nit. Use a constant for this value?

File: libbeat/processors/dissect/dissect_test.go (hunk 9 above)
Comment (jsoriano) [reply libbeat/processors/dissect/dissect_test.go:80]: > Json handles only **number** data type. We need to test for integer, long, float data types also, which is not possible in Json. That's why I could not add test cases to dissect_tests.json and added them to dissect_tests.go instead.

Oh, makes sense :+1:

File: libbeat/processors/dissect/field.go [hunk 13]
```diff
@@ -353,36 +370,35 @@ func newIndirectField(id int, key string, length int) indirectField {
...
 		},
 	}
 }
 
-func extractKeyParts(rawKey string) (key string, ordinal int, length int, greedy bool) {
+func extractKeyParts(rawKey string) (key string, dataType string, ordinal int, length int, greedy bool) {
 	m := suffixRE.FindAllStringSubmatch(rawKey, -1)
 
+	if m[0][1] != "" {
+		parts := strings.Split(m[0][1], dataTypeSeparator)
```
[the remark this answers is not in the document]
Comment (premendrasingh) [reply libbeat/processors/dissect/field.go]: Moved parsing of data type to regular expression in libbeat/processors/dissect/const.go

File: libbeat/processors/dissect/dissect_test.go (hunk 11 above)
Comment (premendrasingh) [reply libbeat/processors/dissect/dissect_test.go]: Moved data type after the keys and suffixes in regular expression and tests.

File: libbeat/processors/dissect/field.go (hunk 12 above)
Comment (premendrasingh) [reply libbeat/processors/dissect/field.go]: Changed to use constant for this value.

File: libbeat/processors/dissect/field.go [hunk 14]
```diff
@@ -299,5 +397,13 @@ func extractKeyParts(rawKey string) (key string, ordinal int, length int, greedy
 		greedy = true
 	}
 
-	return m[0][1], ordinal, length, greedy
+	if m[0][7] != "" {
+		if m[0][8] == "" {
+			dataType = unsetDataType
```
Comment (jsoriano) [inline libbeat/processors/dissect/field.go:402]: Nit. Nothing serious, but I am a bit concerned about this `unsetDataType`. There are two cases here we we might say that we have an "unset" datatype, one when the pipe (`|`) is not used, then the expression is correct but there is no datatype set, the other one when the pipe is used, but datatype is empty, what is incorrect. I think this may cause confusion if future readings of the code.
I wonder if we could directly expect in the regular expression that the datatype is not empty, and then check the validity of the resulting fields in a more similar way to how other things like conflicting fields are checked [here](https://github.com/elastic/beats/pull/18683/files#diff-4376db67dcdbf120310a4ec6e1523f4dL209).
Or maybe another option is to make the `dataType` a pointer to string, so we use `nil` when the datatype is really not set, but the expression is correct (no `|` is used), and the empty string when `|` is used but is empty.

Just some thoughts, no need to change it, as you prefer :slightly_smiling_face:

**Comment by jsoriano** [comment]
jenkins run the tests please

File: libbeat/processors/dissect/field.go [hunk 15]
```diff
@@ -67,15 +63,18 @@ const (
 	IP
 )
 
+var (
+	unsetDataType = "[unset]"
+)
```
[the remark this answers is not in the document]
Comment (premendrasingh) [reply libbeat/processors/dissect/field.go]: Refactored code and removed need for const.

File: libbeat/processors/dissect/field.go (hunk 14 above)
Comment (premendrasingh) [reply libbeat/processors/dissect/field.go]: Thanks for the suggestion, refactored code and added check for suffix |, when data type is missing. Removed need for unset to check for missing data type.

File: libbeat/processors/dissect/field.go [hunk 16]
```diff
@@ -37,11 +41,37 @@ type field interface {
...
+	Long
+	Float
+	Double
+	String
+	Boolean
+	IP
+)
+
+var dataTypeNames = map[string]dataType{
+	"[unset]": unset,
```
Comment (jsoriano) [inline libbeat/processors/dissect/field.go:67]: `unset` wouldn't be needed now

File: libbeat/processors/dissect/field.go [hunk 17]
```diff
@@ -63,6 +93,9 @@ func (f baseField) Length() int {
 func (f baseField) Key() string {
 	return f.key
 }
+func (f baseField) DataType() string {
```
Comment (jsoriano) [inline libbeat/processors/dissect/field.go:96]: Nit. Add a new line before.
```suggestion

func (f baseField) DataType() string {
```

File: libbeat/processors/dissect/field.go (hunk 14 above)
Comment (jsoriano) [reply libbeat/processors/dissect/field.go]: Thanks for this refactor!

File: dev-tools/mage/check.go [hunk 18]
```diff
@@ -99,7 +99,7 @@ func GitDiffIndex() ([]string, error) {
 			return nil, errors.Wrap(err, "failed to dissect git diff-index output")
 		}
 
-		paths := strings.Split(m["paths"], "\t")
+		paths := strings.Split(m["paths"].(string), "\t")
```
Comment (sayden) [inline dev-tools/mage/check.go:102]: Why this change here?

Comment (premendrasingh) [reply dev-tools/mage/check.go:102]: When I run **make fmt** I get following  error because d.Dissect() on line #97 now returns **map[string]interface{}**
`
~/go/src/github.com/elastic/beats/libbeat $ make fmt
github.com/elastic/beats/v7/dev-tools/mage
../dev-tools/mage/check.go:102:27: cannot use m["paths"] (type interface {}) as type string in argument to strings.Split: need type assertion
Error: error compiling magefiles
make: *** [add-headers] Error 1
`

Comment (premendrasingh) [reply dev-tools/mage/check.go:102]: I ran the benchmark tests on **master** branch and **dissect_type_aware** branch. Most of the test cases degrade about 10%. Here is comparison of ns/op for master vs this feature


Test   case | New feature ns/op | Master ns/op | Diff
-- | -- | -- | --
BenchmarkDissect/When_all_the_defined_fields_are_captured_by_we_have_remaining_data-12 | 1253 | 1105 | 13%
BenchmarkDissect/Complex_stack_trace-12 | 1559 | 1359 | 15%
BenchmarkDissect/fails_when_delimiter_is_not_found_at_the_beginning_of_the_string-12 | 0.000022 | 0.000022 | 0%
BenchmarkDissect/fails_when_delimiter_is_not_found_after_the_key-12 | 0.000027 | 0.000026 | 4%
BenchmarkDissect/simple_dissect-12 | 852 | 815 | 5%
BenchmarkDissect/dissect_two_replacement-12 | 952 | 1066 | -11%
BenchmarkDissect/one_level_dissect_not_end_of_string-12 | 916 | 872 | 5%
BenchmarkDissect/one_level_dissect-12 | 886 | 832 | 6%
BenchmarkDissect/multiple_keys_dissect_end_of_string-12 | 953 | 903 | 6%
BenchmarkDissect/multiple_keys_not_end_of_string-12 | 1032 | 960 | 8%
BenchmarkDissect/simple_ordered-12 | 1189 | 1076 | 11%
BenchmarkDissect/simple_append-12 | 1160 | 1078 | 8%
BenchmarkDissect/indirect_field-12 | 1006 | 949 | 6%
BenchmarkDissect/skip_field-12 | 929 | 890 | 4%
BenchmarkDissect/named_skiped_field_with_indirect-12 | 1005 | 959 | 5%
BenchmarkDissect/pointer_field_with_indirect-12 | 997 | 935 | 7%
BenchmarkDissect/missing_fields-12 | 1303 | 1167 | 12%
BenchmarkDissect/ignore_right_padding-12 | 1281 | 1028 | 25%
BenchmarkDissect/padding_on_the_last_key_need_a_delimiter-12 | 1110 | 1040 | 7%
BenchmarkDissect/ignore_left_padding-12 | 1083 | 1027 | 5%
BenchmarkDissect/when_the_delimiters_contains_`{`_and_`}`-12 | 1060 | 1004 | 6%
BenchmarkDissect/simple_fixed_length-12 | 1039 | 956 | 9%
BenchmarkDissect/simple_ordered_and_fixed_length_field-12 | 1191 | 1101 | 8%
BenchmarkDissect/simple_padding_and_fixed_length_field-12 | 1227 | 1111 | 10%
BenchmarkDissect/mixed_pointer_and_indirect_and_fixed_length-12 | 1029 | 961 | 7%
BenchmarkDissect/fails_when_there_is_remaining_string_after_the_fixed-length_key-12 | 0.000052 | 0.000055 | -5%
BenchmarkDissect/fails_when_there_is_no_enough_string_for_the_fixed-length_key-12 | 0.000027 | 0.000026 | 4%
BenchmarkDissect/Regular_expression-12 | 422 | 421 | 0%
BenchmarkDissect/Larger_regular_expression-12 | 2794 | 2793 | 0%
BenchmarkDissect/regular_expression_to_match_end_of_line-12 | 586 | 571 | 3%

**Comment by sayden** [comment]
I have been running quite a lot of local tests with this PR and they worked as expected with all the "parts" using dissect across Beats, but the performance increase is worrying.

Do you think there's a way to improve performance? I'm not asking to do it in this PR, more like "well, maybe we can try or not try this or this" Maybe use a different parser if the `DataType` is detected or not? The problem I see now is that this performance increase affects all instances of dissect, not only the ones using data types

**Comment by vjsamuel** [comment]
@sayden thats a very valid point. should we create a git issue and track the same?

**Comment by vjsamuel** [comment]
@sayden @jsoriano , is there anything else left to address on this one?

**Comment by sayden** [comment]
@vjsamuel sorry for the delay. I think it's better to address the performance issue before merging the PR and potentially release it affecting a lot of running systems. Systems very close to their performance limit (by whatever reason) might have outages after releasing this if they are heavy users of dissect. Dissect is, in general, the recommended approach over grok.

Long story short, dissect should address the performance issues that regular expressions introduce to grok.

So how to you see to try to only use the enhancement of this PR when a datatype is detected and keep the current implementation in master if not? This way, only users using data types will be affected and will be aware in a progressive way (in contrast with upgrading an suddenly having a 8-10% of performance decrease)

How does it sounds? :slightly_smiling_face:

**Comment by premendrasingh** [comment]
@sayden I agree with you on performance concerns.
I have changed to use old code path for dissect and perform data type conversion only when at least 1 field is marked for data conversion. Please review and let me know if you still see any issue with this approach.
Thanks.

**Comment by vjsamuel** [comment]
@sayden any more outstanding concerns? @premendrasingh can you please post updated benchmarks here.

**Comment by premendrasingh** [comment]
Ran the benchmark on master and new branch and took average of 3 runs.


Test   case | New feature ns/op | Master ns/op | Diff
-- | -- | -- | --
BenchmarkDissect/When_all_the_defined_fields_are_captured_by_we_have_remaining_data-12 | 1115 | 1131 | -1%
BenchmarkDissect/Complex_stack_trace-12 | 1451 | 1407 | 3%
BenchmarkDissect/fails_when_delimiter_is_not_found_at_the_beginning_of_the_string-12 | 2.03E-05 | 1.97E-05 | 3%
BenchmarkDissect/fails_when_delimiter_is_not_found_after_the_key-12 | 2.83E-05 | 3.57E-05 | -21%
BenchmarkDissect/simple_dissect-12 | 873 | 882 | -1%
BenchmarkDissect/dissect_two_replacement-12 | 961 | 1024 | -6%
BenchmarkDissect/one_level_dissect_not_end_of_string-12 | 933 | 956 | -2%
BenchmarkDissect/one_level_dissect-12 | 898 | 957 | -6%
BenchmarkDissect/multiple_keys_dissect_end_of_string-12 | 967 | 1018 | -5%
BenchmarkDissect/multiple_keys_not_end_of_string-12 | 1045 | 1072 | -3%
BenchmarkDissect/simple_ordered-12 | 1238 | 1241 | 0%
BenchmarkDissect/simple_append-12 | 1168 | 1148 | 2%
BenchmarkDissect/indirect_field-12 | 1049 | 1004 | 4%
BenchmarkDissect/skip_field-12 | 1053 | 989 | 6%
BenchmarkDissect/named_skiped_field_with_indirect-12 | 1101 | 1082 | 2%
BenchmarkDissect/pointer_field_with_indirect-12 | 1043 | 1085 | -4%
BenchmarkDissect/missing_fields-12 | 1262 | 1329 | -5%
BenchmarkDissect/ignore_right_padding-12 | 1099 | 1085 | 1%
BenchmarkDissect/padding_on_the_last_key_need_a_delimiter-12 | 1123 | 1109 | 1%
BenchmarkDissect/ignore_left_padding-12 | 1125 | 1069 | 5%
BenchmarkDissect/when_the_delimiters_contains_`{`_and_`}`-12 | 1125 | 1070 | 5%
BenchmarkDissect/simple_fixed_length-12 | 1039 | 1027 | 1%
BenchmarkDissect/simple_ordered_and_fixed_length_field-12 | 1201 | 1155 | 4%
BenchmarkDissect/simple_padding_and_fixed_length_field-12 | 1260 | 1172 | 8%
BenchmarkDissect/mixed_pointer_and_indirect_and_fixed_length-12 | 1087 | 1025 | 6%
BenchmarkDissect/fails_when_there_is_remaining_string_after_the_fixed-length_key-12 | 0.000035 | 4.67E-05 | -25%
BenchmarkDissect/fails_when_there_is_no_enough_string_for_the_fixed-length_key-12 | 0.000044 | 0.000035 | 26%
BenchmarkDissect/Regular_expression-12 | 431 | 450 | -4%
BenchmarkDissect/Larger_regular_expression-12 | 2977 | 3173 | -6%
BenchmarkDissect/regular_expression_to_match_end_of_line-12 | 610 | 631 | -3%

**Comment by premendrasingh** [comment]
@sayden Please review the changes. 
Thanks and regards.

**Comment by jsoriano** [comment]
jenkins run the tests please

**Comment by premendrasingh** [comment]
@jsoriano Can you please merge? Errors not related to my changes.
Thanks

**Comment by jsoriano** [comment]
@sayden could you please go on with merge and backport to 7.x if you think this is ready? Thanks!

**Comment by sayden** [comment]
Thank you for contributing!

**Comment by premendrasingh** [comment]
Thank you @sayden, @jsoriano for the guidance.
